### PR TITLE
Accept special chars in resource path of the Open API

### PR DESCRIPTION
**Pull request by renuka-fernando**
### Purpose
Please find the details on the attached issue

### Issues

Fixes #

### Automation tests
 - Unit tests added: Yes
 - Integration tests added: Updated existing tests to support Sem Versions in Rate Limits

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

**Review by malinthaprasan — DISMISSED** [review]

File: adapter/internal/oasparser/envoyconf/envoyconf_internal_test.go [hunk 1]
```diff
@@ -396,6 +396,30 @@ func TestGenerateRegex(t *testing.T) {
...
+		{
+			inputpath:     "/v2/pet/{petId}/images/*",
+			userInputPath: "/v2/pet123/images/123-foo.png",
+			message:       "when the resource ends with * and path params in the middle fails",
+			isMatched:     false,
+		},
+		{
+			inputpath:     "/v2/pet/{petId}/images)/*",
+			userInputPath: "/v2/pet/123/images)/123-foo.png",
+			message:       "when the resource ends with *, special char in the path and path params in the middle passes",
```
Comment (slahirucd7) [inline adapter/internal/oasparser/envoyconf/envoyconf_internal_test.go:414]: Only difference in the line #420 is fails. Is that intended ?
`when the resource ends with *, special char in the path and path params in the middle passes`
`when the resource ends with *, special char in the path and path params in the middle **fails**`

same applies to the line 402 and 408 as well.

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go [hunk 2]
```diff
@@ -1340,19 +1340,25 @@ func generateRoutePaths(xWso2Basepath, basePath, resourcePath string) string {
 
 // generatePathRegexSegment - generates a regex segment for a given resource path
 // resourcePath - resource path of the api
-// options - boolean parameter to indicate whether to generate path segment for substitution string
-func generatePathRegexSegment(resourcePath string, options ...bool) string {
+// skipEscapeMetaChars - skip escaping meta chars (special chars like `.(){}`.) True this for route substitution path.
+func generatePathRegexSegment(resourcePath string, skipEscapeMetaChars bool) string {
```
Comment (slahirucd7) [inline adapter/internal/oasparser/envoyconf/routes_with_clusters.go:1344]: are we expecting to make `skipEscapeMetaChars` as true somewhere in the code ?

Comment (renuka-fernando) [reply adapter/internal/oasparser/envoyconf/routes_with_clusters.go:1344]: Yes, this is set to true when setting the substitution string for the route

File: adapter/internal/oasparser/envoyconf/envoyconf_internal_test.go (hunk 1 above)
Comment (renuka-fernando) [reply adapter/internal/oasparser/envoyconf/envoyconf_internal_test.go:414]: Nope `userInputPath` is different in these two test cases. There is `)` in this path and not in the next one.

**Review by slahirucd7 — APPROVED** [review]

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go (hunk 2 above)
Comment (slahirucd7) [reply adapter/internal/oasparser/envoyconf/routes_with_clusters.go:1344]: Got it. I searched the param only in this PR. Seems it was available previously also.

File: adapter/internal/oasparser/envoyconf/envoyconf_internal_test.go [hunk 3]
```diff
@@ -396,6 +396,36 @@ func TestGenerateRegex(t *testing.T) {
 			message:       "when the resource ends with *, trailing characters substitution passes",
 			isMatched:     false,
 		},
+		{
+			inputpath:     "/v2.0.0/pet/{petId}/images/*",
+			userInputPath: "/v2x0x0/pet/123/images/123-foo.png",
+			message:       "when the version is semver, the dot should not matched with any character",
```
Comment (chathuranga95) [inline adapter/internal/oasparser/envoyconf/envoyconf_internal_test.go:402]: This test case is applicable for all cases. Eg: `2.0.0` and `v2.0`

Shall we Generalize the message? Or we can add 2 separate entries to test semver enabled ones and the older APIs.

**Review by chathuranga95 — DISMISSED** [review]

File: adapter/internal/oasparser/envoyconf/envoyconf_internal_test.go (hunk 3 above)
Comment (renuka-fernando) [reply adapter/internal/oasparser/envoyconf/envoyconf_internal_test.go]: Done with https://github.com/wso2/product-microgateway/pull/3429/commits/eec8652a242a83b98a8caed43825bcb845735744

**Review by chathuranga95 — APPROVED** [review]

**Review by malinthaprasan — APPROVED** [review]

**Review by slahirucd7 — APPROVED** [review]

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(EastUS) cluster : dev-deployment-v2 : 20230905.39

**Comment by choreo-cicd** [comment]
[succeeded]  : dev-deployment-v2 : 20230905.39

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : dev-deployment-v2 : 20230905.39

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : dev-deployment-v2 : 20230905.39

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : prod-deployment-v2 : 20230907.5

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : prod-deployment-v2 : 20230907.5

**Pull request merged**